File: apps/api/src/sentinel_api/api/v1/responses/organization.py
"""Response schemas for organization, workspace, membership, and invitation endpoints."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class UserResponse(BaseModel):
    """Serialized user within organization context."""

    id: str
    email: str
    full_name: str
    avatar_url: Optional[str] = None
    is_active: bool = True


class OrganizationResponse(BaseModel):
    """Serialized organization."""

    id: str
    name: str
    slug: str
    logo_url: Optional[str] = None
    industry: Optional[str] = None
    region: Optional[str] = None
    timezone: Optional[str] = None
    owner_id: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class OrganizationListResponse(BaseModel):
    """Paginated list of organizations."""

    results: List[OrganizationResponse]
    total: int


class WorkspaceResponse(BaseModel):
    """Serialized workspace."""

    id: str
    name: str
    slug: str
    environment: str
    description: Optional[str] = None
    ai_config: Dict[str, Any] = Field(default_factory=dict)
    incident_retention_days: int = 90
    organization_id: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WorkspaceListResponse(BaseModel):
    """List of workspaces within an organization."""

    results: List[WorkspaceResponse]
    total: int


class MembershipResponse(BaseModel):
    """Serialized membership record with embedded user info and scoped workspaces."""

    id: str
    user_id: str
    organization_id: str
    workspace_id: Optional[str] = None
    role: str
    user: Optional[UserResponse] = None
    created_at: Optional[str] = None
    workspaces: List[WorkspaceResponse] = Field(default_factory=list)


class MembershipListResponse(BaseModel):
    """List of memberships within an organization."""

    results: List[MembershipResponse]
    total: int


class InvitationResponse(BaseModel):
    """Serialized invitation."""

    id: str
    email: str
    role: str
    status: str
    organization_id: str
    invited_by: str
    expires_at: Optional[str] = None
    created_at: Optional[str] = None
    workspaces: List[str] = Field(default_factory=list)


class InvitationListResponse(BaseModel):
    """List of invitations within an organization."""

    results: List[InvitationResponse]
    total: int


class MemberActivityResponse(BaseModel):
    """Serialized audit log of member activity."""

    id: str
    organization_id: str
    user_id: Optional[str] = None
    action: str
    details: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None
    user: Optional[UserResponse] = None


class MemberActivityListResponse(BaseModel):
    """Paginated list of team member activity logs."""

    results: List[MemberActivityResponse]
    total: int

File: apps/web/lib/api/types/organization.ts
// ── Multi-Tenant Organization & Workspace Types ──────────────

export type MemberRole = "owner" | "admin" | "engineer" | "viewer";
export type InvitationStatus = "pending" | "accepted" | "expired" | "revoked" | "cancelled" | "rejected";
export type WorkspaceEnvironment = "development" | "staging" | "production";

export interface UserInfo {
  id: string;
  email: string;
  full_name: string;
  avatar_url?: string | null;
  is_active: boolean;
}

// ── Organization ─────────────────────────────────────────────

export interface Organization {
  id: string;
  name: string;
  slug: string;
  logo_url?: string | null;
  industry?: string | null;
  region?: string | null;
  timezone?: string | null;
  owner_id: string;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface OrganizationListResponse {
  results: Organization[];
  total: number;
}

export interface CreateOrganizationRequest {
  name: string;
  slug: string;
  industry?: string;
  region?: string;
  timezone?: string;
}

export interface UpdateOrganizationRequest {
  name?: string;
  logo_url?: string;
  industry?: string;
  region?: string;
  timezone?: string;
}

// ── Workspace ────────────────────────────────────────────────

export interface Workspace {
  id: string;
  name: string;
  slug: string;
  environment: WorkspaceEnvironment;
  description?: string | null;
  ai_config: Record<string, any>;
  incident_retention_days: number;
  organization_id: string;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface WorkspaceListResponse {
  results: Workspace[];
  total: number;
}

export interface CreateWorkspaceRequest {
  name: string;
  slug: string;
  environment?: WorkspaceEnvironment;
  description?: string;
}

export interface UpdateWorkspaceRequest {
  name?: string;
  description?: string;
  environment?: WorkspaceEnvironment;
  ai_config?: Record<string, any>;
  incident_retention_days?: number;
}

// ── Membership ───────────────────────────────────────────────

export interface Membership {
  id: string;
  user_id: string;
  organization_id: string;
  workspace_id?: string | null;
  role: MemberRole;
  user?: UserInfo | null;
  created_at?: string | null;
  workspaces?: Workspace[];
}

export interface MembershipListResponse {
  results: Membership[];
  total: number;
}

export interface AddMemberRequest {
  email: string;
  role?: MemberRole;
}

export interface UpdateMemberRoleRequest {
  role: MemberRole;
}

// ── Invitation ───────────────────────────────────────────────

export interface Invitation {
  id: string;
  email: string;
  role: MemberRole;
  status: InvitationStatus;
  organization_id: string;
  invited_by: string;
  token: string;
  expires_at?: string | null;
  created_at?: string | null;
  workspaces?: string[];
}

export interface InvitationListResponse {
  results: Invitation[];
  total: number;
}

export interface CreateInvitationRequest {
  email: string;
  role?: MemberRole;
  workspaces?: string[];
}

// ── Activity Log ─────────────────────────────────────────────

export interface MemberActivity {
  id: string;
  organization_id: string;
  user_id?: string | null;
  action: string;
  details: Record<string, any>;
  created_at?: string | null;
  user?: UserInfo | null;
}

export interface MemberActivityListResponse {
  results: MemberActivity[];
  total: number;
}
